### PR TITLE
My flsa fixes

### PR DESCRIPTION
**Pull request by ERijck**
Fixes points 1,3 and 4 from #3437. 

More specifically, I have:

- Addressed the FIXMEs. I removed the ones I fixed and answered the ones `ERijck:` I did not fix.
- Fixed points 1,3, and 4. 
- Removed the method to get topic embeddings. It is not logical to have it in the `flsamodel.py`.

Now, we still need to fix point 2. It is unclear how to perform singular value decomposition on a bow representation. However, I suspect you do this in the `lsimodel.py` as well, so I will see how it is done there and rewrite the code.

### CODE REVIEW

File: gensim/models/flsamodel.py [hunk 1]
```diff
@@ -623,7 +675,21 @@ def _create_partition_matrix(
             method='fcm',
         ):
         """
-        Cluster the projected data.
+        Use fuzzy clustering on the projected data.
+
+        The three clustering methods are:
+            1. fcm (Fuzzy C-Means)
+                (http://bitly.ws/zjs6)
```
Comment (piskvorky) [inline gensim/models/flsamodel.py:682]: Please use full canonical URLs. Avoid shorteners and any obfuscation (security risk).

File: gensim/models/flsamodel.py [hunk 2]
```diff
@@ -1434,11 +1320,12 @@ def __init__(
             min_count=1,
             window=5,
             vector_size=20,
-            workers=4,
         ):
 
         self.model = ...   # FIXME what is this?
         self.word_embedding = ...  # FIXME what is this?
+        #ERijck: this way users can do post-analysis. But storing them as an attribute
+        #       is not crucial.
```
Comment (piskvorky) [inline gensim/models/flsamodel.py:1328]: OK, but what does the `...` do?

It's Python's special syntax for ellipsis, or slicing, but I'm not familiar with using it for attribute values.

File: gensim/models/flsamodel.py [hunk 3]
```diff
@@ -1473,18 +1358,15 @@ def get_word_embedding(
...
         """
         self.model = Word2Vec(
             sentences=data,
             vector_size=vector_size,
             window=window,
             min_count=min_count,
-            workers=workers,
         )
         # FIXME is the whole Word2Vec model really needed? Why are we storing it as an attribute?
+        #ERijck: This way users can do a post analysis. However, it is not crucial and can be removed.
```
Comment (piskvorky) [inline gensim/models/flsamodel.py:1369]: OK. Best add a comment to that effect: 

`# Storing the whole trained model is not needed. We're only doing it so users can do a post analysis. Could be removed.`

File: gensim/models/flsamodel.py [hunk 4]
```diff
@@ -359,7 +392,8 @@ def _create_sparse_global_term_weights(
                 word_to_index,
             )
         else:
-            raise ValueError('Invalid word weighting method')
+            raise ValueError(f'Unsupported word_weighting {word_weighting}')
+        #ERijck: This ValueError can be omitted, as it is an internal method. Do you agree?
```
Comment (piskvorky) [inline gensim/models/flsamodel.py:396]: Up to you. In general, input validation should only happen at public API points, where user sends in some data. 

And even there, the "pythonic" way is to rely on duck typing, and not require specific classes or methods.

File: gensim/models/flsamodel.py [hunk 5]
```diff
@@ -955,19 +892,18 @@ def show_topics(
...
             )
         if not isinstance(index_to_word, dict):
-            raise TypeError("Please feed the algorithm 'index_to_word' as a dict")
+            raise TypeError("Please feed a dict to 'index_to_word'")
         if not isinstance(num_words, int) or num_words <= 0:
             raise TypeError("Please use a positive int for 'num_words'.")
         if prob_word_given_topic.shape[0] < prob_word_given_topic.shape[1]:
             raise ValueError("'prob_word_given_topic' has more columns then rows,",
                              " probably you need to take the transpose.")  # FIXME What? Why?
+            #ERijck: A model with more topics than words makes no sense.
```
Comment (piskvorky) [inline gensim/models/flsamodel.py:906]: Maybe, but that's not an error. At most a warning.

**Review by piskvorky — CHANGES_REQUESTED** [review]
This is great, the documentation makes much more sense now.

I left a few minor comments; the only bigger step left is to accept streaming input = avoid materializing the whole corpus in-RAM. I'll try to get to it this weekend, to assist you.

File: gensim/models/flsamodel.py (hunk 2 above)
Comment (mpenkov) [reply gensim/models/flsamodel.py:1328]: It's a built-in ellipsis constant. People use it as a placeholder (TODO).

```
Python 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> x = ...
>>> x
Ellipsis
```

Comment (piskvorky) [reply gensim/models/flsamodel.py:1328]: Yes, that's what I wrote. But why is it here?

Comment (piskvorky) [reply gensim/models/flsamodel.py:1328]: If they're really "required" public attributes with deferred initialization, use simply `self.model = None`. No need for special ellipsis.

Plus maybe add a comment about why these attributes are not initialized right away. Unless it's obvious (sorry I haven't checked the code in detail).

**Comment by ERijck** [comment]
I have just fixed the small comments you left:

image

However, when I push this the same as how I did before, I get the following error:

image

When I try to pull the branch, git seems up to date already:

image

Which commands can I use to push my latest commit: 7864ebc8cbb5f54095b2b90d6b72255c915433c3?

**Comment by piskvorky** [comment]
@ERijck we'll be releasing the hotfix soon, which only removes the `setup.py` dependency. Can you finish up this PR please?

Your error above means you rebased your PR, instead of doing simple commit, so the changes are "non-fast-forward". What did you do exactly? Maybe some graphic git tool messed up your repository. Prefer the command line.

**Comment by ERijck** [comment]
Hi @piskvorky, thanks for your message. I will look into this with a colleague soon.

**Comment by piskvorky** [comment]
@ERijck any plans to finish this?

I'm thinking it's best if you put this (fixed) module into a separate package, separate from gensim. It's useful on its own, but looks like it's not ready for gensim yet. The maintenance effort to support this module seems too steep for us.

**Comment by ERijck** [comment]
@piskvorky, I fixed the points on March 21st. So that only step 2 (avoid materializing the whole corpus in RAM) is left. 
For this, I was waiting for your input to finish, but I understand that this takes too much time and effort for you.

What needs to change so that it is ready for Gensim? Is it correct that you want to use a BOW representation using integers to represent a document rather than a list of strings? If so, I can implement these changes. If you want other changes, could you briefly describe what you want/how I can do that? Then, I will work on these implementations.

**Pull request merged**